feat: implement the editor

reJoker opened the pull request:
## 🤔 Why

This is a take home exam.

## 💡 How

- Handled most of the styles of components with `styled-components` library.
- Get the file data through File API
- Listens to the mouse events to update the rectangles

reJoker committed:
build: add styled-components

reJoker committed:
build: add @types/styled-components

reJoker committed:
feat(App.tsx): init landing page

Replace the generated react landing page with an simple dummy page.

reJoker committed:
feat: init ImagePreviewPane

reJoker committed:
feat(ImagePreviewPane): image select and preview functions

With Files API, we could trigger let user select the files and read the
values.

reJoker committed:
feat(ImagePreviewPane): implement the image selection functions

Listens to mouse events and draws the selection rects.

reJoker committed:
fix(ImagePreviewer): prevent selection in HTML

With the `user-select` property, we could prevent selection of the
ImagePreviewer component while selecting rects.

https://developer.mozilla.org/en-US/docs/Web/CSS/user-select

reJoker committed:
feat(ImagePreviewPane): move selectedRects to App component

We are expecting to pass the value to a sibling component of
ImagePreviewPane. We should hold the state on the parent component.

reJoker committed:
feat(DataPreviewPane): create DataPreviewPane component

reJoker committed:
feat(App): show rect info on data preview pane

reJoker committed:
style(App): change styles of the two panes

reJoker committed:
chore: update page title

reJoker committed:
style(DataPreviewPane): update styles of DataPreviewPane

reJoker committed:
refactor(ImageUploader): move the compoent to independent file

reJoker committed:
style(App): set global style box-sizing

reJoker committed:
style(ImageUploader): update styles of the component

reJoker committed:
refactor(ImagePreviewer): abstract ImagePreviewer component

This component is aimed to display the image and listen to mouse events
to draw rects.

reJoker committed:
style(ImagePreviewPane): update layout

reJoker committed:
refactor: rename `selectedRects` to `ownedRects`

We are going to implement the selecting feature, rename the variables
could reduce confusion.

reJoker committed:
feat(ImagePreviewer): add moving functionality

reJoker committed:
style(ImagePreviewer): visual effect on moving rect

Set a background color to the moving element.

reJoker committed:
chore(ImagePreviewer): update the select rule

It will select the latest created element when there are multiple
elements at the same position.

reJoker committed:
feat(ImagePreviewer): add delete function

reJoker committed:
test: update test case

Joseph7451797 commented on src/components/ImagePreviewer.tsx:
```diff
@@ -0,0 +1,253 @@
+    width: number;
+    height: number;
+    isCompleted: boolean;
+    _id: string;
```
Any specific reason to use `_id` rather than `id` here?

reJoker replied:
I set it to `_id` because I sometimes work on the document from mongo DB.
I will change it to `id`.

Joseph7451797 commented on src/components/ImageUploader.tsx:
```diff
@@ -0,0 +1,73 @@
+        }
+        inputFieldRef.current.click();
+    };
+    const handleImageDataChange = () => {
+        if (!inputFieldRef.current) {
+            return alert("image uploader is not ready");
+        }
+        const fileList = inputFieldRef.current.files;
+        if (!fileList || fileList.length < 1) {
+            return console.error("failed to get the fileList");
+        }
+        const file = fileList.item(0);
+
+        props.setImageData(URL.createObjectURL(file));
+    };
```
Any reason that use ref rather than change event to get file here?

reJoker replied:
I did not recall it could be done in this way.
I will update it to listen to the `onChange` event.

reJoker commented on src/components/ImagePreviewer.tsx:
```diff
@@ -0,0 +1,253 @@
+        }
+
+        let found = false;
+        const nextRects = Array.from(ownedRects)
+            .reverse()
+            .map((rect) => {
+                if (found) {
+                    return rect;
+                }
+                if (
+                    (relativeX - rect.x) * (relativeX - rect.x - rect.width) <
+                        0 &&
+                    (relativeY - rect.y) * (relativeY - rect.y - rect.height) <
+                        0
+                ) {
+                    found = true;
+                    return {
+                        ...rect,
+                        isMoving: true,
+                    };
+                }
+                return rect;
+            })
+            .reverse();
```
1. I would like to select the latest created rectangle and don't want to change the orders of the rectangles. It could also be done by finding the last match and map it again.
2. It's computing if a point p with the coordinate `(relativeX, relativeY)` is inside the rectangle which diagonal is `p1(rect.x, rect.y)` and `p2(rect.x + rect.width, rect.y + rect.height)`. The rectangle is not skewed and not rotated so we could say p is inside the rectangle if p.x is between p1.x and p2.x and p.y is between p1.y and p2.y.
```
// relativeX is between `rect.x` and `rect.x + rect.width`
(relativeX - rect.x) * (relativeX - rect.x - rect.width) < 0
// relativeY is between `rect.y` and `rect.y + rect.height`
(relativeY - rect.y) * (relativeY - rect.y - rect.height) < 0

reJoker committed:
perf(ImagePreviewer): find rect been clicked with for loop

Refactor the expressions of selecting clicked rectangle, using a for
loop from the end of the array instead of reverse the array twice.

Adds comments to describe how we determines if the point is inside a
rectangle.

reJoker committed:
refactor(ImagePreviewer): using `id` instead of `_id`

reJoker committed:
fix(ImageUploader): get file data from change event

reJoker committed:
feat(ImageUploader): checks the MIME type of the uploaded file

Popout alert when user selects a file not is an image.

Joseph7451797 commented:
As the specifications say, the selection blocks should be resizable. Can you enhance them?
